### PR TITLE
Add support for test case folders

### PR DESCRIPTION
**Pull request by stefanius**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -6,7 +6,7 @@
     <groupId>com.testmonitor</groupId>
     <artifactId>api-java-client</artifactId>
     <packaging>jar</packaging>
-    <version>1.4.3</version>
+    <version>1.5.0</version>
```
Comment (thijskok) [inline pom.xml:9]: Should we make this 2.0.0?

File: src/main/java/com/testmonitor/actions/TestCaseFolders.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    public TestCaseFolder create(TestCaseFolder testCaseFolder) throws IOException {
+        JSONObject response = this.connector.post("test-case/folders", testCaseFolder.toHttpParams());
+
+        testCaseFolder.setId(response.getJSONObject("data").get("id").toString());
+
+        return testCaseFolder;
+    }
+
+    /**
+     * Clome a test case folder
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCaseFolders.java:100]: Little typo

File: src/main/java/com/testmonitor/actions/TestCases.java [hunk 3]
```diff
@@ -102,15 +101,15 @@ public ArrayList<TestCase> search(String query) throws IOException, URISyntaxExc
      * Search though test cases in a test suite.
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCases.java:101]: Suite?

File: src/main/java/com/testmonitor/actions/TestCaseFolders.java (hunk 2 above)
Comment (thijskok) [reply src/main/java/com/testmonitor/actions/TestCaseFolders.java:100]: More down below.

File: src/main/java/com/testmonitor/actions/TestCases.java [hunk 4]
```diff
@@ -234,4 +233,36 @@ public TestCase update(TestCase testCase) throws IOException {
...
+
+    /**
+     * Move a test case to another folder
+     *
+     * @param testCaseId The test case id you want to update
+     * @param folderId The test case folder id you want to update
+     *
+     * @return The moved test case
+     */
+    public TestCase move(int testCaseId, int folderId) throws IOException {
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCases.java:245]: We both use testCaseFolderId and folderId, shouldn't we pick one? testCaseFolderId?

File: src/main/java/com/testmonitor/actions/TestCases.java [hunk 5]
```diff
@@ -234,4 +233,36 @@ public TestCase update(TestCase testCase) throws IOException {
 
         return TestCaseParser.parse(updatedTestCase);
     }
+
+    /**
+     * Move a test case to another folder
+     *
+     * @param testCaseId The test case id you want to update
+     * @param folderId The test case folder id you want to update
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCases.java:241]: I think you mean: the destination folder.

File: src/main/java/com/testmonitor/actions/TestCaseFolders.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+        params.add(new BasicNameValuePair("project_id", this.projectId.toString()));
+        params.add(new BasicNameValuePair("page", page.toString()));
+        params.add(new BasicNameValuePair("limit", limit.toString()));
+
+        return TestCaseFolderParser.parse(this.connector.get("test-case/folders", params));
+    }
+
+    /**
+     * @param id The test case ID
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCaseFolders.java:68]: test case ID?

File: src/main/java/com/testmonitor/actions/TestCases.java [hunk 7]
```diff
@@ -136,15 +135,15 @@ public TestCase create(TestCase testCase) throws IOException {
      * Create a test case using the provided name and test suite ID.
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCases.java:135]: test suite?

File: src/main/java/com/testmonitor/actions/TestCaseFolders.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+     * @param includeSubfolders Sets the option to include subfolders
+     * @param includeTestCases Sets the option to include testcases
+     *
+     * @return The cloned test case folder
+     */
+    public TestCaseFolder clone(TestCaseFolder source, String name, boolean includeSubfolders, boolean includeTestCases) throws IOException {
+        List<NameValuePair> params = new ArrayList<>();
+
+        params.add(new BasicNameValuePair("name", name));
+        params.add(new BasicNameValuePair("include_subfolders", Convert.booleanToNumericString(includeSubfolders)));
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCaseFolders.java:113]: Isn't this just:

includeTestCases ? "1" : "0"

It only occurs twice, so...

File: src/main/java/com/testmonitor/actions/TestCases.java [hunk 9]
```diff
@@ -102,15 +101,15 @@ public ArrayList<TestCase> search(String query) throws IOException, URISyntaxExc
...
      *
      * @return A list of results
      */
-    public ArrayList<TestCase> search(String query, Integer testSuiteId) throws IOException, URISyntaxException {
+    public ArrayList<TestCase> search(String query, Integer testCaseFolderId) throws IOException, URISyntaxException {
         List<NameValuePair> params = new ArrayList<>();
 
         params.add(new BasicNameValuePair("project_id", this.projectId.toString()));
-        params.add(new BasicNameValuePair("test_suite", testSuiteId.toString()));
+        params.add(new BasicNameValuePair("test_case_folder_id", testCaseFolderId.toString()));
```
[the remark this answers is not in the document]
Comment (stefanius) [reply src/main/java/com/testmonitor/actions/TestCases.java:112]: @thijskok Yes there is one. But I did not touch it: https://github.com/testmonitor/api-java-client/pull/34/files/f431ad66d22a3e09f9a6eea18aad0d6baef25cec#diff-800fc9febf1ce2fbb073d2070024583877eb718bd9524aa3cf3d0ee4e6c1387bR90

File: src/main/java/com/testmonitor/actions/TestCaseFolders.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    /**
+     * Create a test case folder
+     *
+     * @param testCaseFolder The test case your want to create
+     *
+     * @return The created test case folder
+     */
+    public TestCaseFolder create(TestCaseFolder testCaseFolder) throws IOException {
```
Comment (thijskok) [inline src/main/java/com/testmonitor/actions/TestCaseFolders.java:100]: A proper method for creating a folder nested into an other folder is missing

File: src/main/java/com/testmonitor/resources/TestCaseFolder.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
+package com.testmonitor.resources;
+
+import org.apache.hc.core5.http.NameValuePair;
+import org.apache.hc.core5.http.message.BasicNameValuePair;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestCaseFolder {
```
Comment (thijskok) [inline src/main/java/com/testmonitor/resources/TestCaseFolder.java:9]: A constant defining root would be nice?

File: src/main/java/com/testmonitor/actions/TestCaseFolders.java (hunk 10 above)
Comment (stefanius) [reply src/main/java/com/testmonitor/actions/TestCaseFolders.java:100]: It should be possible to fill the parentId directly on the model, but for some convenience I added a createSubfolder method that will do the trick.

File: src/main/java/com/testmonitor/resources/TestCaseFolder.java (hunk 11 above)
Comment (stefanius) [reply src/main/java/com/testmonitor/resources/TestCaseFolder.java:9]: I added it on the actions

Comment (thijskok) [reply src/main/java/com/testmonitor/resources/TestCaseFolder.java:9]: Actions... is a bit weird? It isn't an action?

TestCaseFolder.root(projectId) makes more sense?

**Review by thijskok — COMMENTED** [review]
See comment

**Review by thijskok — APPROVED** [review]

**Pull request merged**